File: test/resources/repos/polyglot/test_repo/typescript/calculator.ts
// TypeScript Calculator implementation for polyglot testing

export class Calculator {
    value: number;

    constructor(initialValue: number = 0) {
        this.value = initialValue;
    }

    add(x: number): number {
        this.value += x;
        return this.value;
    }

    subtract(x: number): number {
        this.value -= x;
        return this.value;
    }

    multiply(x: number): number {
        this.value *= x;
        return this.value;
    }

    divide(x: number): number {
        if (x === 0) {
            throw new Error("Cannot divide by zero");
        }
        this.value /= x;
        return this.value;
    }

    reset(): number {
        this.value = 0;
        return this.value;
    }
}

export function helperDouble(x: number): number {
    return x * 2;
}

export function helperSquare(x: number): number {
    return x * x;
}

File: test/resources/repos/polyglot/test_repo/haskell/src/Calculator.hs
-- | Haskell Calculator implementation for polyglot testing

module Calculator
    ( Calculator(..)
    , newCalculator
    , add
    , subtract
    , multiply
    , divide
    , reset
    , helperDouble
    , helperSquare
    ) where

import Prelude hiding (subtract, divide)

-- | Calculator data type with current value
data Calculator = Calculator
    { value :: Double
    } deriving (Show, Eq)

-- | Create new calculator with initial value
newCalculator :: Double -> Calculator
newCalculator initialValue = Calculator { value = initialValue }

-- | Add x to current value
add :: Calculator -> Double -> Calculator
add calc x = calc { value = value calc + x }

-- | Subtract x from current value
subtract :: Calculator -> Double -> Calculator
subtract calc x = calc { value = value calc - x }

-- | Multiply current value by x
multiply :: Calculator -> Double -> Calculator
multiply calc x = calc { value = value calc * x }

-- | Divide current value by x
divide :: Calculator -> Double -> Either String Calculator
divide calc 0 = Left "Cannot divide by zero"
divide calc x = Right $ calc { value = value calc / x }

-- | Reset value to zero
reset :: Calculator -> Calculator
reset calc = calc { value = 0 }

-- | Helper function that doubles a number
helperDouble :: Double -> Double
helperDouble x = x * 2

-- | Helper function that squares a number
helperSquare :: Double -> Double
helperSquare x = x * x
